feat: added Caddy proxy setup instructions for Wordpress integration

ksylvan opened the pull request:
### CHANGES

- Added Caddy as a reverse proxy documentation section.
- Provided setup for serving Wordpress alongside Mailu.
- Detailed creating a standalone docker-compose file for Caddy.
- Included instructions to modify `docker-compose.yml` for Caddy usage.
- Explained SSL certificate handling with Let's Encrypt and Caddy.

## What type of PR?

documentation

## What does this PR do?

This PR adds comprehensive documentation for using Caddy as a reverse proxy with Mailu, providing an alternative to the existing Traefik setup. The new documentation explains how to configure Caddy to handle web-facing applications while allowing Mailu to manage mail-related services.

This also avoids the complexity of managing proxy-protocol when using the Traefik as a proxy setup.

### Related issue(s)

- #3643 
- #3052 
- #2700 

## Prerequisites

N/A

nextgens reviewed (dismissed):
- This breaks Mailu's rate-limiter as it masquerades the IP address of the request. At the very least `X-Forwarded-For` should be set and Mailu should be configured to accept it (and that's where things stop being simple).
- This will bind a specific type of SSL certificate (rather than using RSA/ECDSA as appropriate based on client capabilities)
- the software is called Mailu, nothing else.

More generally speaking, I'm not sure we want this at all. There are good reasons why we want SSL termination to be done by Mailu.

ksylvan commented:
> * This breaks Mailu's rate-limiter as it masquerades the IP address of the request. At the very least `X-Forwarded-For` should be set and Mailu should be configured to accept it (and that's where things stop being simple).
> * This will bind a specific type of SSL certificate (rather than using RSA/ECDSA as appropriate based on client capabilities)
> * the software is called Mailu, nothing else.
> 
> More generally speaking, I'm not sure we want this at all. There are good reasons why we want SSL termination to be done by Mailu.

The rate limiter seems to be working as expected. I see lines like this in my logs:

```
admin-1            | [2024-12-18 17:26:41,467] WARNING in limiter: Authentication attempt from 87.120.93.10 has been rate-limited.
```

I'll see what I can figure out about the SSL RSA/ECDSA issue. The SSL certs are still handled by Mailu and served on every port except for https, which

nextgens commented:
> The rate limiter seems to be working as expected.

Try logging in on the web-interface using different, non-existing usernames up to ``AUTH_RATELIMIT_IP`` times (5/hour by default) through the proxy. Odds are this will prevent any login from any IP as the proxy itself will end up in the rate-limit.

It can't work any other way if you have not set ``REAL_IP_FROM``

ksylvan commented:
Thanks, @nextgens for that feedback.

I changed the configuration slightly, and changed the documentation to add `REAL_IP_FROM` and `REAL_IP_HEADER` and now I see this when attempting to do a bad login:

```
admin-1            | [2024-12-18 20:24:04,489] INFO in base: Login attempt for: foo@gmail.com/sso/https from: 67.180.199.110/None: failed: badauth: f4861e
```

instead of:

```
admin-1            | [2024-12-18 20:28:49,776] INFO in base: Login attempt for: foo@gmail.com/sso/https from: 192.168.203.8/47138: failed: badauth: f4861e
```

ksylvan commented:
I updated the setup and it addresses most of the questions.

> Hmm, it's still missing source port information (see the None?).

Yes, and I'm not exactly sure where to fix that. When I do an invalid login, I see this in the admin logs:

```text
admin-1            | [2024-12-22 00:02:03,317] INFO in base: Login attempt for: foo@gmail.com/sso/https from: 67.180.199.110/None: failed: badauth: 67b797
```

In my caddy debug logs, I see the following headers are being set. You can see the `X-Real-IP` and `X-Real-Port` headers being sent.

```json
{
  "remote_ip": "67.180.199.110",
  "remote_port": "55970",
  "client_ip": "67.180.199.110",
  "method": "GET",
  "host": "mail.example.com",
  "headers": {
    "Sec-Fetch-Mode": [
      "cors"
    ],
    "Sec-Fetch-Dest": [
      "font"
    ],
    "X-Real-Ip": [
      "67.180.199.110"
    ],
    "X-Forwarded-Proto": [
      "https"
    ],
    "X-Forwarded-Host": [
      "mail.example.com"
    ],
    "X-Forwarded-For": [
      "67.180.199.110"
    ],
    "User-Agent": [
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
    ],
    "X-Real-Port": [
      "55970"
    ]
  },
  "tls": {
    "resumed": false,
    "version": 772,
    "cipher_suite": 4865,
    "proto": "h2",
    "server_name": "mail.example.com"
  }
}
```

> As you have now realised, it's a lot trickier to get right than it looks.

Yes. 💯 I appreciate the work and thought that has gone into Mailu.

> I have also noticed that what you are proposing will only suggest a single FQDN. This will break client auto configuration, mta-sts, ...

You can do multiple FQDN and mta-sts and auto-configuration (my new update shows all that).

One nit: The default MailU configuration (with or without using an external reversproxy) serves the `apple.mobileconfig` with a `text/html` content type. That causes the page to simply be displayed and no auto-config happens.

I fixed that in my `Caddyfile` by having it change the content type to be:

```
Content-Type application/x-apple-aspen-config
```

And now it behaves as you would expect:

![image](https://github.com/user-attachments/assets/02fd7c6a-2076-4407-bd64-1d0a593dbca0)

nextgens commented:
> One nit: The default MailU configuration (with or without using an external reversproxy) serves the `apple.mobileconfig` with a `text/html` content type. That causes the page to simply be displayed and no auto-config happens.
> 
> I fixed that in my `Caddyfile` by having it change the content type to be:
> 
> ```
> Content-Type application/x-apple-aspen-config
> ```
>

[...]

Good catch!

It's the wrong place where to fix it though. I've submitted #3691

nextgens commented:
> > Hmm, it's still missing source port information (see the None?).
> 
> Yes, and I'm not exactly sure where to fix that. When I do an invalid login, I see this in the admin logs:

I suggest you read http://nginx.org/en/docs/http/ngx_http_realip_module.html

nextgens commented:
`X-Forwarded-For` needs to be formatted as per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986).

`X-Real-Port` and `X-Real-IP` are internal things that we use that caddy should not set

ghostwheel42 commented:
I think adding [X-Forwarded-For](https://github.com/Mailu/Mailu/pull/3684/files#diff-a9bb069c3b466936ce6291330aa58be5be8a4962da65a0df87565062868aea5bR214) will help.
Keeping X-Real-IP and X-Real-Port doesn't hurt, but the backends need to be configured accordingly (like the nginx serving the wordpress for example).

nextgens commented:
agreed

bors try
